Improve reliability and logging of setup node.

evanlinjin opened the pull request:
Fixes #379

 Changes:	
* Split setup node into multiple functions for easier testability.
* Check inputs for each setup request.
* Use context.Context in all parts of the setup process to ensure we cancel on hard timeout.
* Introduced routerclient.Map to clean up various parts of the logic, as well as allow the setup process to be speedier.
* Got rid of the routing.Path data type as it is confusing. I has resorted to use []routing.Hop directly.
* Unit tests.

How to test this PR:
* `make check`
- [x] I need to update `skywire-services` interactive env to work with these changes.

jdknives commented on README.md:
```diff
@@ -111,10 +111,10 @@ $ skywire-cli -h
 After `skywire-node` is up and running with default environment, default apps are run with the configuration specified in `skywire-config.json`. Refer to the following for usage of the default apps:
 
-- [Chat](/cmd/apps/chat)
+- [Chat](/cmd/apps/skychat)
```
Why not rename here and in the cases below?

jdknives commented on Makefile:
```diff
@@ -28,10 +28,10 @@ config: ## Generate skywire.json
-install: ## Install `skywire-node`, `skywire-cli`, `manager-node`, `therealssh-cli`	
-	${OPTS} go install ./cmd/skywire-node ./cmd/skywire-cli ./cmd/setup-node ./cmd/manager-node ./cmd/therealssh-cli	
+install: ## Install `skywire-node`, `skywire-cli`, `manager-node`, `SSH-cli`	
+	${OPTS} go install ./cmd/skywire-node ./cmd/skywire-cli ./cmd/setup-node ./cmd/manager-node ./cmd/therealssh-cli
```
Could you point out why it needs to remain `therealssh-cli` here?

ayuryshev replied:
Two reasons:

1. `go install` has no param allowing customizing name of installed binary (no `-o` flag like in `go build`) - so installed binary will have name of package
2. We can't name golang package as `SSH-cli` - `make lint` will not allow us this

Possible ways:
1. Create installation procedure ourselves: compile, find suitable path, put binary in this path
2. Bend linter

evanlinjin committed:
Improve reliability and logging of setup node.
* Split setup node into multiple functions for easier testability.
* Check inputs for each setup request.
* Use context.Context in all parts of the setup process to ensure we cancel on hard timeout.
* Introduced routerclient.Map to clean up various parts of the logic, as well as allow the setup process to be speedier.
* Got rid of the routing.Path data type as it is confusing. I has resorted to use []routing.Hop directly.
* Unit tests.

evanlinjin committed:
Merge from 'origin/develop'.

nkryuchkov commented on .travis.yml:
```diff
@@ -1,7 +1,7 @@
 go:
   # - "1.13.x" At minimum the code should run make check on the latest two go versions in the default linux environment provided by Travis.
-  - "1.13.x"
+  - "1.14.x"
```
Could you please update the comment?

nkryuchkov commented on pkg/setup/node_old_test.go:
```diff
@@ -0,0 +1,248 @@
+
+package setup
+
+// TODO(evanlinjin): Either fix or rewrite these tests.
```
I think it would be better if these tests are fixed/rewritten before merging.

nkryuchkov commented on pkg/setup/node_test.go:
```diff
@@ -1,276 +1,147 @@
+			rtIDR := newMockReserver(t, nil)
+
+			// act
+			fwd, rev, inter, err1 := GenerateRules(rtIDR, []routing.Route{tc.fwd, tc.rev})
```
Can `err1` be renamed to `err`?

nkryuchkov reviewed (changes requested):
Great improvements!

evanlinjin committed:
Changes as suggested by @nkryuchkov

* Rewrote setup.Node tests.

* Various tweaks/renamings.

evanlinjin committed:
Merge branch 'develop' into fix/setup-node-hangs

evanlinjin committed:
Format code.

nkryuchkov commented on pkg/setup/id_reserver_test.go:
```diff
@@ -80,10 +80,10 @@ func TestIdReserver_ReserveIDs(t *testing.T) {
-		paths    [][]routing.Hop                         // arrange: idReserver input
-		expErr   error                                   // assert: expected error
+		testName string                        // test name
+		routers  map[cipher.PubKey]interface{} // arrange: map of mock router gateways
```
Why would `*mockGatewayForDialer` not be better here?

evanlinjin replied:
@nkryuchkov because we may have different implementations for the RPC Gateway for future tests.

In fact, as I was writing these tests, there was a time point when I had multiple implementations, but I deemed them not necessary.

I think we should leave it as it, just in case for possible future changes.

nkryuchkov replied:
@evanlinjin What do you think about creating a common interface for RPC Gateway implementations and use it in this `map`?

nkryuchkov commented on pkg/setup/node_test.go:
```diff
@@ -17,6 +20,215 @@ import (
+		hash = cipher.SumSHA256(append(pk1[:], pk2[:]...))
+	} else {
+		hash = cipher.SumSHA256(append(pk2[:], pk1[:]...))
+	}
```
Could you please add empty lines before and after `if` block? I think it may improve readability a bit.

nkryuchkov commented on pkg/setup/node_test.go:
```diff
@@ -17,6 +20,215 @@ import (
+	for i, srcPK := range fwdPKs[:len(fwdPKs)-1] {
+		dstPK := fwdPKs[i+1]
+		fwdHops[i] = routing.Hop{TpID: determineTpID(srcPK, dstPK), From: srcPK, To: dstPK}
+	}
```
Could you please add a new line after `for` statement?

nkryuchkov commented on pkg/routing/route.go:
```diff
@@ -76,6 +76,21 @@ func (br *BidirectionalRoute) Check() error {
+		"keep_alive": br.KeepAlive.String(),
+		"fwd_hops":   br.Forward,
+		"rev_hops":   br.Reverse,
+	}
```
Would you mind adding an empty line after this line and before `return` statement?

nkryuchkov reviewed (changes requested):
Great job! Just a few minor requests.

evanlinjin committed:
Tweaks as suggested by @nkryuchkov

Actual dialing to the snettest.Env struct has also been temporarily removed. This will be addressed in #395
